addons/addon-base-raas/packages/base-raas-services/lib/data-source/data-source-reachability-service.js

```diff
@@ -0,0 +1,197 @@
+    // TODO: Determine if the stack is compliant
+    const reachable = false;
+
+    if (reachable) {
+      newStatus = 'reachable';
+      statusMsg = '';
+    } else if (prevStatus === 'pending') {
+      newStatus = prevStatus;
+      statusMsg = `WARN|||Data source account ${id} is not reachable yet`;
+    } else {
+      newStatus = 'error';
+      statusMsg = `ERR|||Error getting information from data source account ${id}`;
+    }
+    if (prevStatus !== newStatus || forceCheck) {
+      const workflowTriggerService = await this.service('workflowTriggerService');
+      await workflowTriggerService.triggerWorkflow(
+        requestContext,
+        { workflowId: workflowIds.accountStatusChange },
+        {
+          id,
+          type,
+        },
+      );
+    }
+    await accountService.updateStatus(requestContext, dataSourceAccount, { status: newStatus, statusMsg });
+    const outputVal = newStatus;
+
+    // Write audit event
+    await this.audit(requestContext, {
+      action: 'check-dsAccount-reachability',
+      body: { id, type },
+    });
+
+    return outputVal;
+  }
+
+  async reachStudy(requestContext, { id, type }) {
+    const studyService = await this.service('studyService');
+    const studyEntity = await studyService.mustFind(requestContext, id);
+
+    if (!studyEntity.status) {
+      throw this.boom.badRequest('Can only check reachability for data source account studies', true);
+    }
+
+    let newStatus;
+    let statusMsg;
+    const prevStatus = studyEntity.status;
+    const reachable = await this._assumeAppRole(studyEntity);
+
+    if (reachable) {
+      newStatus = 'reachable';
+      statusMsg = '';
+    } else if (prevStatus === 'pending') {
+      newStatus = prevStatus;
+      statusMsg = `WARN|||Study ${id} is not reachable yet`;
+    } else {
+      newStatus = 'error';
+      statusMsg = `ERR|||Error getting information from study ${id}`;
+    }
+    await studyService.updateStatus(requestContext, studyEntity, { status: newStatus, statusMsg });
+    const outputVal = newStatus;
+
+    // Write audit event
+    await this.audit(requestContext, {
+      action: 'check-study-reachability',
+      body: { id, type },
+    });
+
+    return outputVal;
+  }
+
+  async _assumeAppRole(studyEntity) {
+    const aws = await this.service('aws');
+    let reachable = false;
+    try {
+      const s3Client = await aws.getClientSdkForRole({
+        roleArn: studyEntity.appRoleArn,
+        clientName: 'S3',
+        options: { region: studyEntity.region },
+      });
+      // use s3Client to read the head of an object
+      await s3Client.headObject({ Bucket: studyEntity.bucket, Key: studyEntity.folder });
+      reachable = true;
+    } catch (err) {
+      // Error is expected if assuming role is not successful yet
+      reachable = false;
+    }
+    return reachable;
+  }
+
+  async attemptReach(requestContext, { id, status, type }, forceCheckAll = false) {
```

Comment:
See my other comment regarding the option object

Reply:
Done